core/src/main/scala/com/salesforce/op/stages/impl/insights/RecordInsightsLOCO.scala

```diff
@@ -158,70 +150,93 @@ class RecordInsightsLOCO[T <: Model[T]]
   private def convertToTimePeriod(descriptorValue: String): Option[TimePeriod] =
     descriptorValue.split("_").lastOption.flatMap(TimePeriod.withNameInsensitiveOption)
 
-  private def getRawFeatureName(history: OpVectorColumnHistory): Option[String] = history.grouping match {
-    case Some(grouping) => history.parentFeatureOrigins.headOption.map(_ + "_" + grouping)
-    case None => history.parentFeatureOrigins.headOption
+  private def getRawFeatureName(history: OpVectorColumnHistory): Option[String] = {
+    val groupSuffix = history.grouping.map("_" + _).getOrElse("")
+    val name = history.parentFeatureOrigins.headOption.map(_ + groupSuffix)
+
+    // If the descriptor value of a derived date feature exists, then it is likely to be
+    // from unit circle transformer. We aggregate such features for each (rawFeatureName, timePeriod).
+    name.map(_ +
+      history.descriptorValue
+        .flatMap(convertToTimePeriod)
+        .map(p => "_" + p.entryName).getOrElse(""))
   }
 
   private def returnTopPosNeg
   (
-    featureArray: Array[(Int, Double)],
+    featureSparse: SparseVector,
+    zeroCountByFeature: Map[String, Int],
     featureSize: Int,
     baseScore: Array[Double],
     k: Int,
     indexToExamine: Int
   ): Seq[LOCOValue] = {
-
     val minMaxHeap = new MinMaxHeap(k)
     val aggregationMap = mutable.Map.empty[String, (Array[Int], Array[Double])]
-    for {i <- featureArray.indices} {
-      val (oldInd, oldVal) = featureArray(i)
-      val diffToExamine = computeDiffs(i, oldInd, oldVal, featureArray, featureSize, baseScore)
-      val history = histories(oldInd)
 
+    agggregateDiffs(featureSparse, indexToExamine, minMaxHeap, aggregationMap,
+      baseScore)
+
+    // Adding LOCO results from aggregation map into heaps
+    for {(name, (indices, ar)) <- aggregationMap} {
+      // The index here is arbitrary
+      val (i, n) = (indices.head, indices.length)
+      val zeroCounts = zeroCountByFeature.get(name).getOrElse(0)
+      val diffToExamine = ar.map(_ / (n + zeroCounts))
+      minMaxHeap enqueue LOCOValue(i, diffToExamine(indexToExamine), diffToExamine)
```

Comment:
wait so we are aggregating everything into a map and then putting it into a heap and then just taking it out of the heap? doesn't that defeat the whole purpose of the heap? Shouldn't we be putting each value into the heap as we calculating it rather than aggregating the whole thing?

Reply:
We are only aggregating TF and Date features

Reply:
ah ok - can you add a comment to that effect